File: source-code/convolution/cpp/src/benchmark_convolution.cpp
#include <convolution.h>
#include <chrono>
#include <iostream>
#include <numeric>
#include <random>

Matrix create_image(int rows, int cols) {
    std::mt19937 gen(1234);
    std::uniform_real_distribution<double> dis(0.0, 1.0);
    Matrix image(rows, cols);
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            image(i, j) = dis(gen);
        }
    }
    return image;
}

Matrix create_kernel(int rows, int cols) {
    Matrix kernel(rows, cols);
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            kernel(i, j) = 1.0/(rows*cols);
        }
    }
    return kernel;
}

double element_sum(const Matrix& matrix) {
    return std::accumulate(matrix.data(), matrix.data() + matrix.rows()*matrix.cols(), 0.0);
}

int main(int argc, char** argv) {
    int rows = 1000;
    int cols = 1000;
    int kernel_rows = 7;
    int kernel_cols = 7;
    if (argc > 1) {
        rows = atoi(argv[1]);
        cols = atoi(argv[1]);
    }
    if (argc > 2) {
        kernel_rows = atoi(argv[2]);
        kernel_cols = atoi(argv[2]);
    }
    std::cout << "Image size: " << rows << "x" << cols << "\n";
    std::cout << "Kernel size: " << kernel_rows << "x" << kernel_cols << "\n";
    Matrix image = create_image(rows, cols);
    Matrix kernel = create_kernel(kernel_rows, kernel_cols);
    auto start = std::chrono::high_resolution_clock::now();
    auto result = convolve(image, kernel);
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end - start;
    std::cout << "Time: " << diff.count() << " s\n";
    std::cout << "Sum: " << element_sum(result) << "\n";
    return 0;
}

File: source-code/convolution/python/benchmark_convolution.py
#!/usr/bin/env python

import argparse
import convolution
import numpy as np
import timeit

def create_image(rows, cols):
    return np.random.uniform(0.0, 1.0, size=(rows, cols))

def create_kernel(rows, cols):
    return np.ones((rows, cols))/(rows*cols)

def element_sum(matrix):
    return np.sum(matrix)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--rows", type=int, default=1_000)
    parser.add_argument("--cols", type=int, default=1_000)
    parser.add_argument("--kernel_rows", type=int, default=7)
    parser.add_argument("--kernel_cols", type=int, default=7)
    args = parser.parse_args()

    print(f'Image size: {args.rows}x{args.cols}')
    print(f'Kernel size: {args.kernel_rows}x{args.kernel_cols}')

    image = create_image(args.rows, args.cols)
    kernel = create_kernel(args.kernel_rows, args.kernel_cols)


    print(f'Time: {timeit.timeit(lambda: convolution.convolve(image, kernel), number=1)} s')
    print(f'Sum: {element_sum(convolution.convolve(image, kernel))}')

if __name__ == "__main__":
    main()
